File: modules/ghidra/ghidra_scripts/ImportChanges.java
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileOptions;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.decompiler.flatapi.FlatDecompilerAPI;
import ghidra.app.script.GhidraScript;
import ghidra.app.script.GhidraState;
import ghidra.framework.model.Project;
import ghidra.framework.model.ProjectLocator;
import ghidra.program.flatapi.FlatProgramAPI;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Variable;
import ghidra.program.model.pcode.HighFunction;
import ghidra.program.model.pcode.HighFunctionDBUtil;
import ghidra.program.model.pcode.HighSymbol;
import ghidra.program.model.pcode.LocalSymbolMap;
import ghidra.program.model.symbol.SourceType;
import ghidra.program.model.symbol.Symbol;
import ghidra.util.exception.DuplicateNameException;
import ghidra.util.exception.InvalidInputException;
import ghidra.util.task.ConsoleTaskMonitor;


import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class ImportChanges extends GhidraScript {

    @Override
    public void run() throws Exception {
		FlatProgramAPI fpapi = new FlatProgramAPI(getState().getCurrentProgram());
		FlatDecompilerAPI fdapi = new FlatDecompilerAPI(fpapi);
		DecompileOptions options = new DecompileOptions();
		ConsoleTaskMonitor monitor = new ConsoleTaskMonitor();
		DecompInterface ifc = new DecompInterface();
		ifc.setOptions(options);
		ifc.openProgram(getState().getCurrentProgram());
		String[] args = getScriptArgs();
		if (args.length > 0) {
			importChanges(args[0], fpapi, ifc, monitor);
		} else {
			importChanges(fpapi, ifc, monitor);
		}
	}

	public void importChanges(FlatProgramAPI fpapi, DecompInterface ifc, ConsoleTaskMonitor monitor) throws InvalidInputException, DuplicateNameException, IOException {
		importChanges(null, fpapi, ifc, monitor);
	}

	public void importChanges(String js_fp, FlatProgramAPI fpapi, DecompInterface ifc, ConsoleTaskMonitor monitor) throws IOException, InvalidInputException, DuplicateNameException {
		GhidraState state = getState();
		Project project = state.getProject();
		ProjectLocator locator = project.getProjectData().getProjectLocator();
		FunctionManager fm = currentProgram.getFunctionManager();
		ArrayList<Function> funcs = new ArrayList<>();
		fm.getFunctions(true).forEachRemaining(funcs::add);

		String programe_name = fpapi.getCurrentProgram().toString().split(" ")[0].replace(".", "_");
		if (js_fp == null) {
			//js_fp = Paths.get(Pathing.PROJECTS_ROOT.toString(), programe_name, programe_name+".json").toString();
			;
		} else {
		    js_fp = Paths.get(js_fp, programe_name+".json").toString();
		}


		String read = Files.readString(Paths.get(js_fp));
		JsonObject save_file = new Gson().fromJson(read, JsonObject.class);
		JsonObject functions_dict;
		if (save_file.keySet().contains("functions")) {
			functions_dict = save_file.getAsJsonObject("functions");
		} else {
			functions_dict = save_file;
		}

		HashMap<String, String> current_lookup = new HashMap<>();
		HashMap<String, String> original_lookup = new HashMap<>();
		for (String function_name : functions_dict.keySet()) {
			JsonObject data = functions_dict.getAsJsonObject(function_name);
			String current_name = data.get("current_name").getAsString();
			current_lookup.put(function_name, current_name);
			original_lookup.put(current_name, function_name);
		}

		for (Function func : funcs) {
			String func_name = func.getName();

			if (!current_lookup.containsKey(func_name)) {
				if (original_lookup.containsKey(func_name)) {
					func_name = original_lookup.get(func_name);
				}
			}

			if (functions_dict.has(func_name)) {
				JsonObject function_data = functions_dict.getAsJsonObject(func_name);
				JsonObject renaming_dict = function_data.getAsJsonObject("renaming");
				String new_name = function_data.get("current_name").getAsString();

				if (function_data.get("skipped").getAsBoolean() | function_data.has("imported") & function_data.get("imported").getAsBoolean()) {
					continue;
				}

				if (!function_data.get("improved").getAsBoolean()) {
					continue;
				}

				if (!func.getName().equals(new_name)) {
					func.setName(new_name, SourceType.IMPORTED);
					System.out.println("Renaming" + func_name + " to " + new_name);
				}

				HighFunction hf = get_high_function(func, ifc, monitor);
				Iterator<HighSymbol> symbols = get_function_symbols(func, ifc, monitor);

				for (Iterator<HighSymbol> it = symbols; it.hasNext(); ) {
					HighSymbol high_symbol = it.next();
					String symname = high_symbol.getName();
					if (renaming_dict.has(symname)) {
						new_name = renaming_dict.get(symname).getAsString();
						if (new_name.equals("") | new_name.equals(symname) | new_name.contains(" ") | new_name.contains(",")) {
							continue;
						}
						Symbol symbol = high_symbol.getSymbol();
						if (symbol == null) {
							continue;
						}
						try {
							symbol.setName(new_name, SourceType.IMPORTED);
						} catch (DuplicateNameException dne) {
							new_name = new_name + "_";
							try {
								symbol.setName(new_name, SourceType.IMPORTED);
							} catch (DuplicateNameException dne2) {
								continue;
							}
						} catch (Exception e) {
							if (e.toString().contains("NoneType")) {
								continue;
							}
							System.out.println("Error while renaming " + symname + " in function " + func_name);
							System.out.println(e.toString());
							continue;
						}
						}
					HighFunctionDBUtil.commitLocalNamesToDatabase(hf, SourceType.IMPORTED);
					HighFunctionDBUtil.commitParamsToDatabase(hf, true, SourceType.IMPORTED);

					ArrayList<Variable> vars_params = new ArrayList<>(Arrays.stream(func.getAllVariables()).toList());
					vars_params.addAll(Arrays.stream(func.getParameters()).toList());
					for (Variable var : vars_params) {
						String var_name = var.getName();
						if (renaming_dict.has(var_name)) {
							new_name = renaming_dict.get(var_name).getAsString();
							if (new_name.equals("") | new_name.equals(var_name) | new_name.contains(" ") | new_name.contains(",")) {
								continue;
							}
							try {
								var.setName(new_name, SourceType.IMPORTED);
							} catch (DuplicateNameException dn) {
								new_name = new_name  + "_";
								try {
									var.setName(new_name, SourceType.IMPORTED);
								} catch (DuplicateNameException dn2) {
									continue;
								}
							}
						}
					}
					functions_dict.get(func_name).getAsJsonObject().addProperty("imported", true);
				}
			}
		}

		for (String func_name : functions_dict.keySet()) {
			JsonObject data = functions_dict.getAsJsonObject(func_name);
			if (!data.get("imported").getAsBoolean()) {
				functions_dict.getAsJsonObject(func_name).addProperty("imported", true);
			}
		}

		if (save_file.keySet().contains("functions")) {
			save_file.add("functions", functions_dict);
		} else {
			save_file = functions_dict;
		}
		Gson gson = new GsonBuilder().setPrettyPrinting().create();
		String pretty_sf = gson.toJson(save_file);
		Files.write(Paths.get(js_fp), pretty_sf.getBytes());
	}

		public HighFunction get_high_function(Function func, DecompInterface ifc, ConsoleTaskMonitor monitor) {
			DecompileResults res = ifc.decompileFunction(func, 60, monitor);
			HighFunction high = res.getHighFunction();
			return high;
		}

		public Iterator<HighSymbol> get_function_symbols (Function func, DecompInterface ifc, ConsoleTaskMonitor monitor) {
			HighFunction hf = get_high_function(func, ifc, monitor);
			LocalSymbolMap lsm = hf.getLocalSymbolMap();
			return lsm.getSymbols();
		}
}

File: modules/ghidra/ghidra_scripts/ImportChanges.py
# loads changes from a file and applies them to the current program
# @author MrMatch246
# @category tooling
# @keybinding
# @menupath Tools.ImportChanges
# @toolbar
import json
import sys
from ghidra.app.decompiler.flatapi import FlatDecompilerAPI
from ghidra.program.flatapi import FlatProgramAPI
import ghidra.program.model.listing.Function
from ghidra.util.task import ConsoleTaskMonitor
from ghidra.app.decompiler import DecompileOptions, DecompInterface
from ghidra.program.model.pcode import HighFunctionDBUtil
from pathing import *
import ghidra.program.model.symbol.SourceType.IMPORTED as IMPORTED

# As we import changes from a file, we mark them as imported and not as user defined

try:
    from ghidra.ghidra_builtins import *
except:
    pass

fpapi = FlatProgramAPI(getState().getCurrentProgram())
fdapi = FlatDecompilerAPI(fpapi)
options = DecompileOptions()
monitor = ConsoleTaskMonitor()
ifc = DecompInterface()
ifc.setOptions(options)
ifc.openProgram(getState().getCurrentProgram())

def main(json_file_path=None):
    state = getState()
    project = state.getProject()
    locator = project.getProjectData().getProjectLocator()
    fm = currentProgram.getFunctionManager()
    funcs = list(fm.getFunctions(True))  # True means 'forward'

    program_name = str(fpapi.getCurrentProgram()).split(" ")[0].replace(".", "_")
    if json_file_path is None:
        json_file_path = os.path.join(
            PROJECTS_ROOT, program_name, f"{program_name}.json"
        )
    else:
        json_file_path = os.path.join(json_file_path, f"{program_name}.json")

    save_file = {}
    with open(json_file_path, "r") as f:
        save_file = json.load(f)
        if "functions" in save_file.keys():
            functions_dict = save_file["functions"]
        else:
            functions_dict = save_file

    current_lookup = {}
    original_lookup = {}
    # Lookup tables
    for function_name, data in functions_dict.items():
        current_lookup[function_name] = data["current_name"]
        original_lookup[data["current_name"]] = function_name

    for func in funcs:

        # Getting original function name if it already has been in ghidra
        func_name = str(func.getName())
        if func_name not in current_lookup.keys():
            if func_name in original_lookup:
                func_name = original_lookup[func_name]

        # Renaming if it is present in the json
        if func_name in functions_dict:
            function_data = functions_dict[func_name]
            renaming_dict = function_data["renaming"]
            new_name = function_data["current_name"]

            # We can skip functions that are Skipped in the json
            if function_data["skipped"] or "imported" in function_data.keys() and function_data["imported"]:
                continue

            # We can skip functions that are not improved in the json (This is not really needed but might be useful)
            if not function_data["improved"]:
                continue

            # Renaming the function
            if func.getName() != new_name:
                func.setName(new_name, IMPORTED)
                print(f"Renaming {func_name} to {new_name}")
            # print("Symbols:")

            # Getting symbols and HighFunction
            # Symbols contain register variables besides other things
            hf = get_high_function(func)
            symbols = get_function_symbols(func)

            for high_symbol in symbols:
                # print(high_symbol.getName())
                symname = high_symbol.getName()
                if symname in renaming_dict.keys():
                    new_name = renaming_dict[symname]
                    if new_name == "" or new_name == symname or " " in new_name or "," in new_name:
                        continue
                    symbol = high_symbol.getSymbol()
                    if symbol is None:
                        continue
                    try:
                        symbol.setName(new_name, IMPORTED)


                    except ghidra.util.exception.DuplicateNameException:
                        new_name = f"{new_name}_"
                        try:
                            symbol.setName(new_name, IMPORTED)
                        except ghidra.util.exception.DuplicateNameException:
                            continue

                    except Exception as e:
                        if "NoneType" in str(e):
                            continue
                        print(f"Error while renaming {symname} in function {func_name}")
                        print(e)
                        continue
                                    #print("Symbol Renaming " + symname + " to " + new_name + " in function " + func_name)

            # Committing changes to the database
            HighFunctionDBUtil.commitLocalNamesToDatabase(hf, IMPORTED)
            HighFunctionDBUtil.commitParamsToDatabase(hf, True, IMPORTED)

            # print("Variables:")

            # Getting variables and parameters of the normal function object
            vars_params = []
            vars_params += func.getAllVariables()
            vars_params += func.getParameters()
            vars_params = list(dict.fromkeys(vars_params))
            for var in vars_params:
                var_name = var.getName()
                if var_name in renaming_dict.keys():
                    # print(var.getName())
                    new_name = renaming_dict[var_name]
                    if new_name == "" or new_name == var_name or " " in new_name or "," in new_name:
                        continue
                    try:
                        var.setName(new_name, IMPORTED)
                    except ghidra.util.exception.DuplicateNameException:
                        new_name = f"{new_name}_"
                        try:
                            var.setName(new_name, IMPORTED)
                        except ghidra.util.exception.DuplicateNameException:
                            continue

                                    #print(str(type(var)) + " Renaming " + var_name + " to " + new_name + " in function " + func_name)
                                    #print("Var Renaming " + var_name + " to " + new_name + " in function " + func_name)
            functions_dict[func_name]["imported"] = True

    for func_name, data in functions_dict.items():
        if not data["imported"]:
            functions_dict[func_name]["imported"] = True

    with open(json_file_path, "w") as f:
        if "functions" in save_file.keys():
            save_file["functions"] = functions_dict
        else:
            save_file = functions_dict
        f.write(json.dumps(save_file, indent=4))


def get_high_function(func):
    res = ifc.decompileFunction(func, 60, monitor)
    return res.getHighFunction()


def get_function_symbols(func):
    hf = get_high_function(func)
    lsm = hf.getLocalSymbolMap()
    return lsm.getSymbols()


if __name__ == "__main__":
    if args := list(getScriptArgs()):
        main(str(args[0]))
    else:
        main()
